### DISCUSSION TITLE
Getting Tokio TCP working with WASM

### DISCUSSION THREAD

**Discussion by maxwellflitton** [tokio-rs/tokio#6526] [Q&A]
Hey I'm the author of Rust and Web Programming for Packt and the O'Reilly book Async Rust. I am trying to map a knowledgebase of WASM in rust to see if I can have a comprehensive guide to networking in WASM. I've managed to get TCP working with `wasmedge` with their own forked libraries. However, now that Tokio unstable supports `wasmtime`, directly running wasm compiled Rust programs is a lot cleaner. So far the following code compiles:

```rust
use std::net::TcpListener as StDTcpListener;
use tokio::net::TcpListener;

#[tokio::main(flavor="current_thread")]
async fn main() {
    use std::os::wasi::io::FromRawFd;
    let std_listener = unsafe { StDTcpListener::from_raw_fd(3) };
    std_listener.set_nonblocking(true).unwrap();
    let listener = TcpListener::from_std(std_listener).unwrap();
    println!("Server listening on port 8080");
}
``` 
With the following dependencies:

```toml
[dependencies]
tokio = { version = "1.37.0", features = ["macros", "io-util", "rt", "time", "net"] }
```

If we use the following commands for the compilation:

```rust
export RUSTFLAGS="--cfg tokio_unstable"
cargo build --target wasm32-wasi --release
cp target/wasm32-wasi/release/wasi-server.wasm wasi-server.wasm
```

We then have the `wasi-server.wasm` file in the root of the project. We can then try and run the file with the following command:

```
export FD_COUNT="3"
wasmtime run ./wasi-server.wasm -W tcp-listen=127.0.0.1:8080
```

However, I get the following error:

```
called `Result::unwrap()` on an `Err` value: Os { code: 8, kind: Uncategorized, message: "Bad file descriptor" }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Error: failed to run main module `./wasi-server.wasm`

Caused by:
    0: failed to invoke command default
    1: error while executing at wasm backtrace:
           0: 0x12317 - <unknown>!__rust_start_panic
           1: 0x12039 - <unknown>!rust_panic
           2: 0x11f6c - <unknown>!std::panicking::rust_panic_with_hook::h39e3a8d718bdeee9
           3: 0x11347 - <unknown>!std::panicking::begin_panic_handler::{{closure}}::he926e6c0f7818bbb
           4: 0x11272 - <unknown>!std::sys_common::backtrace::__rust_end_short_backtrace::h8c8c9b1b5bff6dcb
           5: 0x11943 - <unknown>!rust_begin_unwind
           6: 0x16e8e - <unknown>!core::panicking::panic_fmt::h113f422b14ac6c5e
           7: 0x1892a - <unknown>!core::result::unwrap_failed::h39cc7ddeefacf4e6
           8: 0x2509 - <unknown>!<core::pin::Pin<P> as core::future::future::Future>::poll::hafd7cce1e3bb831b
           9:  0x9d3 - <unknown>!tokio::runtime::scheduler::current_thread::Context::enter::h6571e16d8eddbc1b
          10: 0x237d - <unknown>!tokio::runtime::context::scoped::Scoped<T>::set::hd5c0c2879f3c3696
          11:  0xbb9 - <unknown>!tokio::runtime::scheduler::current_thread::CoreGuard::block_on::h530f9f4b35b42418
          12: 0x2065 - <unknown>!tokio::runtime::context::runtime::enter_runtime::h8de416387b833c99
          13:  0xea2 - <unknown>!wasi_server::main::h9a3299a258d097bb
          14: 0x2252 - <unknown>!std::sys_common::backtrace::__rust_begin_short_backtrace::h08ca4841d5529c37
          15: 0x2261 - <unknown>!std::rt::lang_start::{{closure}}::h64de60455dc43b4c
          16: 0xe919 - <unknown>!std::rt::lang_start_internal::h96857314850acb81
          17: 0x112b - <unknown>!__main_void
          18:  0x5bc - <unknown>!_start
       note: using the `WASMTIME_BACKTRACE_DETAILS=1` environment variable may show more debugging information
    2: wasm trap: wasm `unreachable` instruction executed
```
This happens on the `std_listener.set_nonblocking(true).unwrap();` which suggests that the TCP connection didn't really work and anything happening on the TCP connection causes an error. Does anyone have any references on how to get a TCP connection working with `wasmtime`? I have been burning days looking at repo examples and searching issues in github repos and only found some light references to code that is essentially what we have in this post.

**Comment by Darksonn** [comment]
I'm not familiar with wasm, but that error means that 3 is not a file descriptor. You are using `FD_COUNT=3`, but perhaps that means the file descriptors are 0, 1, and 2? Have you tried passing some other integers to `from_raw_fd`?

**Comment by FrankReh** [comment]
The way to make that work with `wasmtime` at this time is
```sh
wasmtime run -S preview2=n -S tcplisten=127.0.0.1:8080 ./wasi-server.wasm
```

The `FD_COUNT` environment variable doesn't seem to be used by `wasmtime`. I think you can remove it from your setup instructions.

**Comment by maxwellflitton** [comment]
@FrankReh I decided to build out macros that directly manipulate the linear memory in WASM to transfer complex data structures over the WASM boundary. Then it just routes to different functions. You define your kernel for both client and WASM binary like so:

```rust
use nanoservices_utils::create_contract_handler;
use nanoservices_utils::errors::{NanoServiceError, NanoServiceErrorStatus};
use serde::{Deserialize, Serialize};

#[derive(Debug, PartialEq, Serialize, Deserialize)]
pub struct ContractOne {
    pub name: String,
    pub age: u32,
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
pub struct ContractTwo {
    pub account_name: String,
    pub amount: u32,
}

create_contract_handler!(
    ContractHandler, 
    ContractOne, 
    ContractTwo
);
```
https://github.com/nanoservicesforge/nanoservices-utils/blob/wasm-networking/tests/wasm/kernel/src/lib.rs

You then define the routes in the WASM build with code like the following:

```rust
use nanoservices_utils::errors::{NanoServiceError, NanoServiceErrorStatus};
use kernel::{
    ContractHandler,
    ContractOne,
    ContractTwo,
};
use nanoservices_utils::register_wasm_contract_routes;
use paste::paste;

fn handle_contract_one(mut contract: ContractOne) -> Result<ContractOne, NanoServiceError> {
    contract.name = "Bob".to_string();
    Ok(contract)
}

fn handle_contract_two(contract: ContractTwo) -> Result<ContractTwo, NanoServiceError> {
    Ok(contract)
}

register_wasm_contract_routes!(
    ContractHandler,
    handle_contract_routes,
    ContractOne => handle_contract_one,
    ContractTwo => handle_contract_two
);
```
When you build it the macro will generate a load of entry points for the WASM binary and also generates the `malloc` function as I am using raw pointers in the linear memory. The client code is still big because I have not abstracted it yet but the link below shows how your rust code can send a data contract over the WASM boundary and get a response:
https://github.com/nanoservicesforge/nanoservices-utils/blob/wasm-networking/tests/wasm/client/src/main.rs

@Darksonn I have been building an open source package manager on top of Docker so you can declare your docker images in your `Cargo.toml` file and the package manager will download and unpack the code, linking it to your build. I'm a big fan of Tokio framing so I use this for TCP messages between Rust servers with a macro handling the routing:
https://www.nanoservices.io/docs/Docs/iomessaging/receiving-messages
I'm thinking of building a `TcpProxy` actor that listens to a port for these messages and then spawns a task directly interfacing with an instance of the WASM through the client code approach. This means we could have multiple WASM servers being directly interacted with using Tokio TCP. Do you think this is a good idea? Also, there is a lot of mess with async in WASM. I keep getting a load of JS promises dependencies. Sadly I think the functions in the WASM code need to be blocking. Do you know of any approach where I can expose an blocking DB function to the WASM through the host, but have some sort of async functionality in the host? The developer must be able to call the DB function as and when they want. I guess I could just start with a blocking wrapper around the async db call and then try and get something a little more async later on.

**Comment by FrankReh** [comment]
I don't want to be a wet blanket but you might be on your own if you are trying to build something sophisticated using the Wasm core module approach. Wasm work over the past year or more that I'm aware of is focusing on using the new Wasm Component Model. Although the new Wasm C-M is currently released as WASI Preview 2, which does not support async APIs well, there are many in the ByteCode Alliance that are working towards a WASI Preview 3 that will allow things like first class async calls into a Wasm Component image and then it will be possible to run an async runtime like Tokio in a Wasm embedder with one or more Wasm Components, running in separate threads, and composed running in a single thread.

It might still take a year, from what I've read about their efforts, but it is pretty clear (to me) that Wasm Components are going to be the future. (If you want to drink the cool-aid too, search for WebAssembly Component Model on YouTube and see the discussions that have come out over the last year and half, especially the latest ones from early this year.)

Going down the Wasm Component Model route means no longer having to hand craft the glue between the Wasm host and the Wasm guest. The interfaces will now be defined by the Wasm C-M IDL (they call it WIT), and the ABI that the host and guest then must follow is specifically defined and the binding code (the code that used to be called the glue code) is now generated by the build tools.

Admittedly many companies already ship product with the old Wasm Core module idea, successfully, and they will not drop what they have already built, and perhaps some will continue to try and innovate using the old APIs. But I think you find a smaller and smaller audience for that over time. The strongest reason to continue using the Wasm Core module APIs seem to be when targeting browsers or embedded devices that are already in the field and can take Wasm module updates but not Wasm embedder updates.

**Reply by maxwellflitton** [reply]
It's something to look out for, this is why these conversations are useful. The glue code itself can be swapped out, but the routing mimics the same interface as the TCP. So looking at it, I would use wit as my kernel?

**Comment by FrankReh** [comment]
Not sure what you mean by "kernel". WIT is the name they have given to the interface definition language. If a host process, aka an "embedder", can support what the guest Wasm Component module calls out as its interface which is defined by the WIT that comes with the guest, then the host was probably built to support what the guest needs as it makes calls into the guest.

**Reply by maxwellflitton** [reply]
Ah caught up from my domain-driven design reading. It just means something that traverses boundaries to keep a data type of language etc consistent:
https://deviq.com/domain-driven-design/shared-kernel
I'll keep with the Rust kernel for now as it will also enable direct messages to be sent over HTTP/TCP through to the WASM and back again and even propagate the error. I'm interested in building a nice rust ecosystem where people building out servers, clients, and WASM packages in Rust will have smooth experience. This came from managing a kubernetes cluster entirely written in Rust.
